theon/src/x86_64/pc/boot.S

```diff
@@ -412,23 +433,69 @@ eapstart:
 	movw	%ax, %gs
 	movw	%ax, %ss
 
-	movq	APPERCPU, %rdi
-	movq	APID, %rsi
-	movq	APSEMAPHORE, %rdx
+	// We need our x2APIC ID to proceed
+	movl	$IA32_APIC_BASE_MSR, %ecx
+	rdmsr
+	testl	$ApicBaseEnable, %eax
+	jnz	2f
+1:
+	lea	EXApicDisabled(%rip), %rdi
+	hlt
+	jmp	1b
```

Comment:
Maybe leave a note to explain this now weirdly placed jmp statement?

Reply:
I've introduced a label `apearlypanic` with an explanatory comment, and jmp there instead of the loop. I also tidied up a bit more of the assembler there. PTAL